add shellcheck to CI

akuzm opened the pull request:
In my experience, writing a shell script correctly is hard even for a skilled programmer.  [shellcheck](https://github.com/koalaman/shellcheck) is a static analysis tool that helps catch common errors in shell scripts. I found it to be very useful when working with shell scripts.

We now have 36 executable scripts in our repository, for which shellcheck reports 126 errors (calculated like `find . -type f -executable -exec bash -c '[ "$(file --brief --mime-type "$1")" == "text/x-shellscript" ]' sh {} \; -exec shellcheck -f gcc --exclude=SC2086 {} \; | cut -d: -f1 | sort | uniq | wc -l`) I think it might make sense to run shellcheck on changed scripts in pull requests. This PR adds a GitHub actions workflow that does this.

This check only runs on the changed files, so this PR doesn't fix all the existing shellcheck warnings. To test the check, I first fixing a single warning in `test/pg_regress.sh` and confirmed that the check complains about the remaining warnings. I fixed them as well and confirmed that the check passes.

The warning [SC2086: Double quote to prevent globbing and word splitting](https://github.com/koalaman/shellcheck/wiki/SC2086) is disabled globally, because it has little practical consequences, sometimes leads to false positives, and is general is too widespread because people forget to quote.

akuzm commented on .github/workflows/shellcheck.yaml:
```diff
@@ -0,0 +1,35 @@
+
+    - name: Run shellcheck
+      run: |
+        set +e
```
```suggestion
```

svenklemm commented:
I think if we add this it should run on all scripts and all scripts should pass the tests otherwise it will get triggered on unrelated changes to scripts not conforming.

akuzm commented:
> all scripts should pass the tests otherwise it will get triggered on unrelated changes to scripts not conforming.

True, but I was hoping to avoid the heroic effort of fixing everything now by gradually fixing it in the future. I found that shellcheck can correct some things in an automated fashion. I've pushed the result right now. (`find * -type f -executable -exec bash -c '[ "$(file --brief --mime-type "$1")" == "text/x-shellscript" ]' sh {} \; -exec shellcheck -f diff --exclude=SC2086 {} \; | git apply`). I'll review the rest of the errors the next day.

> it should run on all scripts

We can do that, but running only on the diff is cheaper and it prevents introducing new warnings. The only thing it misses is when shellcheck is upgraded and finds some new warnings. The "all files" approach will fail right away, and the "changed files" will fail when the files are modified. Not sure which is preferrable.

akuzm commented:
I fixed all the warnings we currently have, and changed it to always check all the executable shell files in repository.
Not sure what I broke in the scripts that are not ran in CI, maybe I have to test something manually...

afiskon reviewed (approved):
@akuzm thanks for the effort! I think we can give it a try 👍 It's always nice to run some extra static analyzers. The risks seem to be minimal. In the worst-case scenario, the PR can always be reverted.

akuzm commented:
> test bgw_launcher                 ... FAILED     2893 ms

Apparently flaky.

pmwkaa reviewed (approved):
Looks good

akuzm commented on .github/workflows/shellcheck.yaml:
```diff
@@ -0,0 +1,29 @@
+          -print \
+          | grep . \
+          && exit 1 \
+          || exit 0
```
Done.
